src/sql/base/browser/ui/chart/chart.component.ts

```diff
@@ -0,0 +1,319 @@
+
+	public chart: chartjs.Chart;
+	private canvas: HTMLCanvasElement;
+	public element: string;
+
+	/**
+	 * Options in the form that Chart.js accepts (hence the `any` type)
+	 */
+	private _options: any = {
+		events: ['click', 'keyup'],
+		responsive: true,
+		maintainAspectRatio: false
+	};
+
+	constructor(
+		@Inject(forwardRef(() => ChangeDetectorRef)) private _changeRef: ChangeDetectorRef
+	) {
+		chartjs.Chart.register(...chartjs.registerables);
+		super();
+	}
+
+	ngAfterViewInit(): void {
+
+	}
+
+	/**
+	 * Setter function for chart ID
+	 */
+	public set chartId(val: string) {
+		this._chartId = val;
+
+		this.element = this._chartId;
+		this._changeRef.detectChanges();
+	}
+
+	/**
+	 * Setter function for chart type
+	 */
+	public set type(val: any) {
+		if (val === 'horizontalBar') {
+			this._type = 'bar';
+			this._options = mixin({}, mixin(this._options, { indexAxis: 'y' }));
+		}
+		else {
+			this._type = val;
+		}
+		this._changeRef.detectChanges();
+	}
+
+	/**
+	 * Setter function for chart data
+	 */
+	public set data(val: TData) {
+		this._data = this.convert(val);
+
+		this.drawChart();
+	}
+
+	/**
+	 * Setter function for chart options.
+	 * Some options like responsiveness and maintainaspectratio are set by default and will be used even if no options are provided.
+	 */
+	public set options(val: TOptions) {
+		if (val === undefined) {
+			return;
+		}
+
+		// Free-form chart.js options get added first...
+		this._options = mixin({}, mixin(this._options, val.freeformOptions));
+
+		// ...then strongly-typed ComponentModel options get set (overriding free-form options)
+		if (val !== undefined) {
+			if (val.chartTitle) { // undefined results in hiding title
+				if (typeof val.chartTitle === 'string') {
+					this._options = mixin(this._options, {
+						plugins: {
+							title: {
+								text: val.chartTitle,
+								display: true
+							}
+						}
+					});
+				}
+			} else {
+				this._options = mixin(this._options, { plugins: { title: { display: false } } });
+			}
+
+			if (val.legendVisible !== false) { // undefined defaults to true
+				this._options = mixin(this._options, { plugins: { legend: { display: true } } });
+			} else {
+				this._options = mixin(this._options, { plugins: { legend: { display: false } } });
+			}
+		}
+
+		this.drawChart();
+	}
+
+	public set height(val: string | number) {
+		if (val && this.chart) {
+			(this.chart.canvas.parentNode as any).style.height = val;
+		}
+	}
+
+	public set width(val: string | number) {
+		if (val && this.chart) {
+			(this.chart.canvas.parentNode as any).style.width = val;
+		}
+	}
+
+	private convert(val: azdata.ChartData<TChartType>): chartjs.ChartData {
+		const result: chartjs.ChartData = {
+			datasets: []
+		}
+
+		switch (this._type) {
+			case 'bar':
+			case 'horizontalBar': // should've been replaced with 'bar' by this point, but inlcuded case here for safety
+			case 'line':
+				{
+					const data = <azdata.BarChartData><unknown>val;
```

Comment:
Directly casting like this is dangerous, it'd be better to use a [type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards) on the data object IMO (then you could also validate the data passed in is of the correct type

Reply:
A type predicate may not work in this case... Typescript isn't accepting my pinky-swear that `TData` can totally be converted to (e.g.) `BarChartData`, and it'll accept the predicate only if it recognizes the conversion from the predicate type (BarChartData) to whatever TData is (which TS isn't narrowing according to the ChartType param).

![image](https://github.com/microsoft/azuredatastudio/assets/1609827/fd04ecec-edc4-489f-af31-31878e7fcafd)

Plus, the generic types should have already blocked incompatible types (unless someone is creating their own implementations of Data and Options interfaces for an existing chart type).

Maybe I'm missing a way to do this?

Reply:
Have the param be unknown (which is what you'd typically want anyways - since the idea is you don't know what the actual object type is and are trying to find it out)

Reply:
And note you'd have a separate implementation for each type of data as well, then use those in a series of if statements to handle the different scenarios.

Reply:
Dang, yup - I was hoping I'd be able to merge them into a function with a generic parameter, but Javascript's lack of types strikes again.

In any case, added type predicates.